### ISSUE TITLE
RFC - Secure Services by default

### ISSUE THREAD

**Issue by cannikin** [redwoodjs/redwood#1630]
As discussed in the 2021-11-12 Core Meeting, we want to do what we can to make services secure by default. Currently, if you don't add something like `requireAuth()` manually in a services function, any GraphQL endpoints that resolve to those services will happily execute by anyone on the internet. That could be bad.

This issue will serve as a single place to discuss different ways to make services secure by default. Here "by default" means that if you did not do anything special to explicitly allow access, your service/endpoint would be secure—return an error when attempting to access with no authorization/authentication provided.

## Current State

Given the following service function:

```javascript
export const deleteUser = ({ id }) => {
  return db.user.delete({
    where: { id },
  })
}
```

And the following GraphQL SDL:

```graphql
export const schema = gql`
  type User {
    id: Int!
    name: String!
    email: String!
  }

  type Mutation {
    deleteUser(id: Int!): User!
  }
```

Anyone on the internet could perform this mutation and delete every user in the database:

```javascript
import { request, gql } from 'graphql-request'

const query = gql`
  mutation DeleteUser($id: Int!) {
    deleteUser(id: $id) {
      id
    }
  }
`

for (let i=0; i<100000; i++) {
  try {
    await request('https://insecuresite.com/.netlify/functions/graphql', query, { id: i })
  } catch (e) {
    console.log(`User ${i} not found, trying next...`)
  }
}
```

Note that a service can be used internally as well, and the same security implications apply.

This can be prevented by simply adding a `requireAuth()` call inside the service to check for proper permission (either being logged in at all, or the user having a specific role):

```javascript
export const deleteUser = ({ id }) => {
  requireAuth({ roles: ['admin'] })

  return db.user.delete({
    where: { id },
  })
}
```

However, having to add it manually is rife for human error. We want Redwood to be secure by default and help save users from themselves.

## Proposal A

Use some form of `requireAuth()` behind the scenes and add a new function like `allowAccess()` to explicitly allow access.

Assuming the same service:

```javascript
export const deleteUser = ({ id }) => {
  return db.user.delete({
    where: { id },
  })
}
```

Any access to this service would raise an error. You would need to explicitly allow access:

```javascript
export const deleteUser = ({ id }) => {
  allowAccess({ roles: ['admin'] })
  return db.user.delete({
    where: { id },
  })
}
```

Perhaps just calling `allowAccess()` (without arguments) allows any logged in user to access.

Some downsides to this proposal:

* Before adding `allowAccess()` calls, it's not immediately clear that any kind of authorization is happening at all, these just appear to be regular functions
* Not backwards compatible with existing codebases

## Proposal B

This proposal addresses some of the concerns in Proposal A. This solution would be backwards compatible (since it uses the same `requireAuth()` function we already know and love). We add a convention of exporting a function that will automatically run before all services so you only need to declare `requireAuth()` once:

```javascript
export const beforeService = () => {
  requireAuth({ roles: ['admin'] })
}

export const deleteUser = ({ id }) => {
  return db.user.delete({
    where: { id },
  })
}
```

If you have multiple service functions and want to not require auth for one or more of them:

```javascript
export const beforeService = () => {
  requireAuth({ roles: ['admin'] }, except: ['listUsers'])
}

export const listUsers = () => {
  return db.user.findMany()
}

export const deleteUser = ({ id }) => {
  return db.user.delete({
    where: { id },
  })
}
```

Or the opposite where you only want to lock down a few functions:

```javascript
export const beforeService = () => {
  requireAuth({ roles: ['admin'] }, only: ['deleteUser'])
}

export const listUsers = () => {
  return db.user.findMany()
}

export const deleteUser = ({ id }) => {
  return db.user.delete({
    where: { id },
  })
}
```

This proposal has the added benefit of letting you execute additional code that you always want to occur before a group of services (logging the API call, perhaps):

```javascript
import { logAccess } from 'src/services/logger'

export const beforeService = ({ context, serviceName }) => {
  requireAuth({ roles: ['admin'] }, only: ['deleteUser'])
  logAccess(context.currentUser, serviceName)
}

export const deleteUser = ({ id }) => {
  return db.user.delete({
    where: { id },
  })
}
```

What if the service does not export a `beforeService` function? We have a few options:

* Raise an error: if you decide you do not need authorization in your service then you need to make it an explicit, conscious decision by exporting a `beforeService()` that returns nothing
* Not require a `beforeService()` call, but throw an error or at least a warning if we do not encounter a `requireAuth()` function call in a service
* Require an explicit function call like `skipAuth()` call to really make it clear what you're doing

## Conclusion

Please add a comment if you have any additional syntax ideas!

**Comment by mojombo** [comment]
I like a lot of what I see in Option 2.

* Being backwards compatible is pretty nice.
* It allows you to cover a whole service very easily and still be quite visible that auth is happening if you keep that export at the top of the file.
* If each of your "resolver" functions does something totally different, you can decompose the `requireAuth` into each one and leave your `beforeService` empty.

I'd probably expect a missing `beforeService` not to error and just disallow access to everything.

One question here: how does it apply to non-"resolver" functions in a service that are called from other places in the service or other services?

I think a good way to explore this further will be to think about what changes we'd need to make to the tutorial.

**Comment by dthyresson** [comment]
> One question here: how does it apply to non-"resolver" functions in a service that are called from other places in the service or other services?

This is a very good question.

To be honest, I want my resolvers to be secure, but let's say I write a background job (and secure that in whatever propepr way there is) and I wish to call a createModel.

If that createModel service requiresAuth, then I cannot really reuse the service since my background function won't be authenticated.

> think about what changes we'd need to make to the tutorial.

Ideally, the change(s) would be simple, quick, and clear.

The Guild shows that GraphQL directives can define the auth:

https://www.graphql-tools.com/docs/schema-directives/#enforcing-access-permissions

```
directive @auth(
  requires: Role = ADMIN,
) on OBJECT | FIELD_DEFINITION

enum Role {
  ADMIN
  REVIEWER
  USER
  UNKNOWN
}

type User @auth(requires: USER) {
  name: String
  banned: Boolean @auth(requires: ADMIN)
  canPost: Boolean @auth(requires: REVIEWER)
}
```

What I like about this approach is that the policy is centralized, defined, and the service generators can recreate the auth -- such that if one `rw g sdl Model --crud --force` to refresh the model's sdl and services then the auth can be recreated.

For the tutorial, if *you don't* need an authenticated Comment ... don't specify the auth directives -- and when you do, do.

A concern I do have though, is that when generating the sdl now, the directives would be overwritten.

I wonder if the directives could be set in a AuthPolicy.js or ModelAuthPolicy.js or ModelPermission.js such that the are separate and would not be overwritten by incremental model/sdl/service generation.

**Comment by AntonioMeireles** [comment]
@cannikin 

FWIW, at sight **B** also looks from looks like a smoother, less disruptive evolutionary path, as it seems to achieve all of the same goals as **A** (i.e. it seems there's nothing that could be achieved by **A** that couldn't  via **B**) 

@dthyresson, @mojombo  

> I wonder if the directives could be set in a AuthPolicy.js or ModelAuthPolicy.js or ModelPermission.js such that the are separate and would not be overwritten by incremental model/sdl/service generation.

here, imho, one could either follow a similar approach as in `Routes.js` where generators would then populate a given [centralized] file with the generated bits or alternatively have a per Model (paired with each one) a separate file where the {Auth,Model}{Policy,Permissions}   would be set. One could also look at adding to the `rw g sdl Model` cli a new `--auth` option so that when `rw g sdl Model --crud --force` got called auth bits would be preserved and would **only* be overwritten  if `rw g sdl Model --crud --force --auth` is invocated.

**Comment by peterp** [comment]
> One question here: how does it apply to non-"resolver" functions in a service that are called from other places in the service or other services?

Maybe a person could let us know by exporting a `skipBefore` method:

```js
export const numOfActiveUsers = () => {}

export const before = () => {
   requireAuth()
}

export const skipBefore = () => [numOfActiveUsers]
```

**Comment by Tobbe** [comment]
@mojombo 
> * Being backwards compatible is pretty nice.

Agreed

> I'd probably expect a missing `beforeService` not to error and just disallow access to everything.

But then it wouldn't be backwards compatible anymore, right? Everyone's services would suddenly block all access.

**Comment by Tobbe** [comment]
> > One question here: how does it apply to non-"resolver" functions in a service that are called from other places in the service or other services?
> 
> Maybe a person could let us know by exporting a `skipBefore` array of functions:

Or just put those functions in /lib and include them?

**Comment by cannikin** [comment]
I've made some progress!

In the most basic usage:

```javascript
export const before = () => {
  requireAuth({ role: 'moderator' })
}
```

You can also destructure the name of the service that `before()` is about to run for and check that:

```javascript
export const before = ({ serviceName }) => {
  if (serviceName === 'deleteComment') {
    requireAuth({ role: 'moderator' })
  }
}
```

You can also return `false` from `before()` which, right now, returns a "Service aborted" error:

```javascript
export const before = ({ serviceName }) => {
  if (serviceName === 'deleteComment') {
    return false
  }
}
```

This could be handy for situations where you want/need to stop a service from being executed that aren't auth related.

I like @peterp's idea of how to have `before()` skip certain service functions...in my original code snippet this was an option passed to `requireAuth()` but that doesn't really make any sense in retrospect!

**Comment by cannikin** [comment]
Thanks to @peterp's help `before()` now works in Jest! 

We also added `after()` and the ability to send `context` into both `before()` and `after()` so you can inspect `currentUser`, for example:

```javascript
export const before = () => {
  requireAuth({ role: 'moderator' })
}

// not the most performant way to get the results for a single user, but you get the idea
export const after = ({ response }, { context }) => {
  return response.filter((row) => {
    return row.userId === context.currentUser.id
  })
}

export const comments = ({ postId }) => {
  return db.comment.findMany({ where: { postId } })
}
```

@mojombo we should talk more about what happens if you don't provide any `before()` function. You said:

> I'd probably expect a missing `beforeService` not to error and just disallow access to everything.

But right now disallowing access (like with `requireAuth()`) *does* raise an error, `AuthenticationError`. You could return `null` or `[]` but then how do you know the difference between the service being locked down and there just not being any records to return?

There's a `ForbiddenError` that comes with Apollo that we aren't using for anything. We could throw that instead? I'm sure there's a whole discussion to be had about an AuthError vs. a ForbiddenError but that's an HTTP thing and GraphQL doesn't care about that!!

![image](https://user-images.githubusercontent.com/300/104504728-c21c7680-5597-11eb-913b-3de53743dd30.png)

**Comment by cannikin** [comment]
> One question here: how does it apply to non-"resolver" functions in a service that are called from other places in the service or other services?

The way this currently works is that the stuff in `before()` only comes into play when the services are called via GraphQL endpoint. By importing and executing a service function directly you're only invoking that function. Is this good or bad? Do you *want* stuff in `before()` to come into play?

I was thinking of this implementation similar to before_actions in controllers in Rails: it locks down access from the outside world, but if you want to use some piece of code internally in your app there's nothing stopping you from doing so. That means if you put your `requireAuth()` in the service function itself, all calls to that service will be subject to that check, via GraphQL or internally. But if you put it in `before()` then it only applies to GraphQL calls.

Maybe `before()` should be something more like `beforeResolve()` or `beforeResolution()` to make it clear it only comes into play when GraphQL resolves?

We don't have any concept of ActiveRecord lifecycle callbacks (before_save, after_save)...maybe that's the kind of thing that should live in Services and access permissions should live somewhere else? But the GraphQL layer knows nothing about business logic and permissions, is it right for it to live there? hmmm 

I don't love this syntax:

```graphql
directive @auth(
  requires: Role = ADMIN,
) on OBJECT | FIELD_DEFINITION

enum Role {
  ADMIN
  REVIEWER
  USER
  UNKNOWN
}
```

`on OBJECT | FIELD_DEFINITION`? Ugh. And it requires having yet another list of roles stored somewhere. 

@dthyresson In your createModel example, are you doing anything like that already with pyplyn (might have spelled that wrong)? It seems like you'd need to add a header or something to the background job request and check for the presence of that before falling back to `requireAuth`? Can that be accomplished cleanly in a `before()` or in the service itself?

**Comment by cannikin** [comment]
There would be a slight change to how tests are implemented using the code changes we've figured out so far... rather than importing service functions and using them directly, we're providing a top level `services` object that wraps them in the code necessary to enable `before` and `after`:

```javascript
const { comments, deleteComment } = window.services.comments

scenario(
  'returns all comments for a single post from the database',
  async (scenario) => {
    const result = await comments({ postId: scenario.comment.jane.postId })
    expect(result.length).toEqual(1)
  }
)
```

**Comment by papaponmx** [comment]
I like proposal B with the `except: ['listUsers'])` implementation. 
I'd rather have our resolvers over-protected by default instead of having them under-protected.

**Comment by Tobbe** [comment]
Reading through this again I really like where this is going.

For the web side we have `beforeQuery`, so I think `beforeResolve` or `beforeResolution` would go very well with that. It also makes it clearer that the function isn't called before all service calls, just when the services are used as resolvers.

If no `beforeResolution` is exported and no  `requireAuth()` is executed I think we should print a warning. Unless the function is listed in `skipBefore`.

If we throw an error we lose backwards compatibility, which I think would be nice if we could keep. We can always say in the warning that we're planing to make it an error in the future. Also refusing to run the function (even without throwing an error) would break backwards compatibility. 

If the user really doesn't want auth, `export const beforeResolution = () => {}` would be a way to signal that. Or listing the function in `skipBefore`.  `skipAuth()` could be a nice convenience method to signal the same intent. Or maybe adding `unsecure` to the filename `posts.unsecure.js`.

**Comment by cannikin** [comment]
Hmm, do we only want the secure stuff to happen before being used as a resolver though? I was under the impression that ALL code calling these services (other services) should also fall under the security requirements. @mojombo ?

**Comment by Tobbe** [comment]
Sure, but where would you put that implementation though? If I just include a service function from some random place, how would RW know to call `before` before my code runs the function?

**Comment by cannikin** [comment]
This is usually when I shrug my shoulders and say "babel?" and then look over at @peterp

**Comment by cannikin** [comment]
Or maybe it's like this code change shown here: https://github.com/redwoodjs/redwood/issues/1630#issuecomment-760368504  You use the services defined on `window.services` rather than importing them directly. That way we can get in the middle and include a call to `before()` in front of your actual service function `users()` and export *that*.

**Comment by cannikin** [comment]
Okay, after a brainstorming session with @mojombo here's the interface we came up with:

```javascript
export const beforeResolver = (before) => {
  const requireAuthor = () => {
    requireAuth({ roles: ['author'] })
  }

  before.run(requireAuth)
  before.run(requireAuthor, { only: ['createPost'] })
}

export const posts = () => {
  return db.post.findMany()
}

export const createPost = ({ input }) => {
  return db.post.create({ data: input })
}
```

In the `beforeResolver()` function, we're building a "spec" (the `before` object) with all the checks we want to happen before we allow the requested resolver function (in this example, `posts()`) to be invoked. You call `run()` on the `before` object, once for each function you want to invoke. Any of those functions can then throw an error if whatever conditions are not met.

Redwood can ensure that `run()` is called at least once. If it's not, then we throw an error saying you need to explicitly allow access to some services by giving at least one function to `run()`. (We can't actually enforce that the function given to `run()` is doing something auth-related, but we're doing what we can!)

What if an endpoint doesn't require auth, and really *can* be open to the world? You let us know that by by calling `skip()` on the `before` object:

```javascript
export const beforeResolver = (before) => {
  before.skip()
}

export const posts = () => {
  return db.post.findMany()
}
```

Every resolver would need to be explicitly covered by at least one call to `run` or `skip`. That means that a GraphQL call to `posts` would error here, because only `createPosts` is covered by the calls to `run()`:

```javascript
export const beforeResolver = (before) => {
  before.run(requireAuth, { only: ['createPosts'] })
}

export const posts = () => {
  return db.post.findMany()
}

export const createPost = ({ input }) => {
  return db.post.create({ data: input })
}
```

One way to fix that:

```javascript
export const beforeResolver = (before) => {
  before.skip()
  before.run(requireAuth, { only: ['createPosts'] })
}

export const posts = () => {
  return db.post.findMany()
}

export const createPost = ({ input }) => {
  return db.post.create({ data: input })
}
```

Or:

```javascript
export const beforeResolver = (before) => {
  before.run(requireAuth)
  before.skip({ only: ['posts'] })
}

export const posts = () => {
  return db.post.findMany()
}

export const createPost = ({ input }) => {
  return db.post.create({ data: input })
}
```

Note that using `only` and `except` with a `skip` can get kind of confusing (like using a `!` around a conditional check) but it is possible and allows for the most flexibility when defining your spec. We've been using a similar syntax in Rails for years and you get used to it. :) Generally you have the "most common" check be the blanket one for everything in the service, and then use `skip` on the few that don't need it (or vice-versa, if it's more common that most do NOT need anything to happen before).

## Resolving Resolvers

The calls to `run` are "additive", meaning that we start from the first call to `run` to the last before determining which service functions are covered by which `run` calls:

```javascript
export const beforeResolver = (before) => {
  before.run(requireAuth)
  before.run(requireAuthor, { only: ['createPost'] })
  before.skip(requireAuth, { only: ['posts'] })
}
```

1. every service function is covered by `requireAuth`
2. add `requireAuthor` to `createPosts`
3. remove `requireAuth` from `posts`

So the final service-function-to-beforeResolver-function-lookup ends up lb:

```javascript
{ 
  'posts': [], 
  'createPost': ['requireAuth', 'requireAuthor'] 
}
```

## Options

`run()` takes an optional second argument, an object containing option that, for now, are `only` and `except`, which are the names of the services that the `run()` applies to. For example:

```javascript
export const beforeResolver = (before) => {
  before.skip()
  before.run(requireAuth, { except: ['posts'] })
}

export const posts = () => {
  return db.post.findMany()
}

export const createPosts = ({ input }) => {
  return db.post.create({ data: input })
}
```

So here `requireAuth()` is only invoked when trying to access the `createPosts()` service, but not the `posts()` service. If you had more functions you wanted to *ex*clude than *in*clude, you can use the `only` option instead:

```javascript
export const beforeResolver = (before) => {
  before.skip()
  before.run(requireAuth, { only: ['createPosts'] })
}

export const posts = () => {
  return db.post.findMany()
}

export const createPosts = ({ input }) => {
  return db.post.create({ data: input })
}
```

If the majority of your functions didn't need auth, or needed more complex functionality that you don't want to live in the `beforeResolver()` you could call `skip()` everything and manually include your auth checks in the individual service functions:

```javascript
export const beforeResolver = (before) => {
  before.skip()
}

export const posts = () => {
  return db.post.findMany()
}

export const createPost = ({ input }) => {
  requireAuth()
  return db.post.create({ data: input })
}
```

`skip()` also accepts a function as the first parameter, and means "only skip this one function". It also accepts `only` and `except`:

```javascript
export const beforeResolver = (before) => {
  before.run(requireAuth)
  before.run(logRequest)

  before.skip({ only: ['posts'] })                 // skips all resolver functions
  // or
  before.skip(requireAuth, { only: ['posts'] })  // only skips `requireAuth`
}
```

## Only Auth?

Although all of the examples above discuss auth, you could use the spec to include any code you want to run before your service is called via GraphQL:

```
export const beforeResolver = (before) => {
  before.run(requireAuth)
  before.run(circularQueryCheck)
  before.run(allowUserAgent)
  before.run(logRequest)
}
```

Since the functions invoked by `run()` are also on the API side, they'll have access to `context` just like the services themselves do, so you can check the currentUser, inspect the request headers, etc.

## Questions

* Should we make it so that the only way to exit the chain of `run()` calls is if one of them `throw`s? Or should we do something like if `false` is ever returned from one we halt the chain and return a certain error?
* Coming from Ruby land, I would make `run()` super flexible and let you pass multiple functions all at once, and let Redwood worry about whether you're passing a function or an object with options:

```javascript
export const beforeResolver = (before) => {
  before.run(requireAuth, circularQueryCheck, allowUserAgent, logRequest)
  before.skip(requireAuth, only: ['posts'])
}
```

Does that scare JS developers though? :) There's nothing stopping you from writing each on their own line, but this lets you get pretty expressive in fewer lines of code, but isn't hugely obfuscating.

**Comment by Tobbe** [comment]
Are we going to require everyone who upgrades to add `beforeResolver` to all their existing code?

**Comment by cannikin** [comment]
I come down on the "break anything you want before v1.0" side! 

But, maybe it can be an experimental flag you set in `redwood.toml` or something that says whether to enforce that requirement. If you *don't* have the flag, it won't be enforced. But all new apps will be created with the flag enabled, and then in v1.0 the behavior will be that it's *always* be enabled.

**Comment by mojombo** [comment]
@cannikin I like that approach, give people some time to get the code in but still be able to upgrade. Definitely we'll want you to have it by 1.0.

**Comment by dthyresson** [comment]
@cannikin One of the features I liked about GraphQL shied was that it was very easy to "deny all" and then selectively 
"allow".

```js
const permissions = shield({
  Query: {
    "*": deny
    query1: allow,
    query2: allow,
  },
  Mutation: {
    "*": deny
  },
}, {
  fallbackRule: allow
})
```

The philosophy here is to lock down the entire GraphQL API and just allow particular resolvers based on "some rules" that check auth, roles, etc.

` "*": deny` ie deny all

To achieve a "deny", would one implement:

```js
before.run(deny)
```

where `deny()` simply returns `false`?

Or effectively, have something the opposite of `skip` like `before.deny`?

I could very much envision a scenario where I may have some services, but perhaps I want them to me private and *never* exposed to the api. Maybe I use them in a function -- or use them via other services. I'd want to deny their being invoked over the GraphQL API.

TLDR;

* Consider a means to `deny` via a !skip like method easily that RW provides
* Consider a means to do that globally for *all* services/resolvers
* Consider a`before.allow` as an alias for `before.skip` (ie opposite of deny)

**Comment by thedavidprice** [comment]
Assigned this to you @dthyresson and prioritizing for review during the next sprint. Correct assignment and sound good?

**Comment by dthyresson** [comment]
@thedavidprice 

> Correct assignment?

Not sure this is on my task list - this is different that looking into GraphQL security. Those are two separate efforts.

I can review the RFC and make comments, but I don't know if this is the desired implementation yet. Has this been ok'd?

**Comment by thedavidprice** [comment]
Got it. Removing assignment. But keep a priority for discussion at the start of the next cycle. Thanks!

**Comment by cannikin** [comment]
There's really three separate things we've talked about securing:

1. Services (just the services themselves, not the interface to the outside world via GraphQL) - covered by this issue
2. GraphQL endpoints - not sure if there's an open issue/RFC for this one?
3. Custom Functions - sort of discussed in https://github.com/redwoodjs/redwood/issues/1618

**Comment by dthyresson** [comment]
> GraphQL endpoints - not sure if there's an open issue/RFC for this one?

@cannikin and @thedavidprice There isn't *yet* but I have started documenting scenarios (Info disclosure, Depth, Injection, Validation, etc).

**Comment by cannikin** [comment]
Okay, I have an implementation of the "spec" class which is what the `run()` and `skip()` functions are run against! But I have absolutely no idea how to integrate it into Redwood itself, and do things like have GraphQL invoke this functionality before trying to call the actual resolver. I didn't even attempt to put this into the real Redwood codebase. Here's a gist with the BeforeSpec class: https://gist.github.com/cannikin/b8a84601b120d8ba1a1ad93812040022#file-beforespec-js The other files in there are a test suite and a mini assertions lib I wrote (I wasn't even going to attempt to worry about adding webpack/babel/Jest and whatever else would be needed, so I made my own). There's also an `example_service.js` so we can see what's required for a user to get this working in a service.

So, integrating this into Redwood itself somehow...I'm theorizing it would look something like:

```javascript
import * as services from 'src/services/posts'

// user must define/export a `beforeResolver()` function in their service
if (typeof services.beforeResolver === 'undefined') {
  throw new Error('Must define a `beforeResolver()` in services')
}

// get the names of all the services, minus the `beforeResolver`
serviceNames = Object.keys(services).filter(n => n !== 'beforeResolver')

// initialize an instance of the `BeforeSpec` with the names of the services
spec = new BeforeSpec(serviceNames)

// call the `beforeResolver` defined in the service, passing in the spec instance
services.beforeResolver(spec)

/*
** All of the `run()` and `skip()` functions have been called and now we know which functions
** to run before any of the service calls
*/

/*
** Now somewhere/somehow Apollo wants to invoke a service function as a resolver.
** We would intercept that and do something like:
*/

// spec.verify() either:
//  1. Throws an error if no `run()` calls were made for that service, or explicitly skipped with `skip()`
//  2. Bubbles up any thrown error inside one of the functions given to `run()`
//  3. Returns `true`
spec.verify(userInvokedServiceName))

// if we get the actual service function is okay to be called
// TODO: not sure if this should really call with `this` or some scope provided by Apollo
services[userInvokedServiceName].call(this)  
```

I'm still not sure if the only way to get out of a before function should be to `throw`, or if simply returning `false` from any of them should halt the chain and we throw our own error, like `Service validation failed`. Any opinions either way?

@peterp If you're available anytime soon I'd love to do some PAIRING on this!! Let me know what you think!

**Comment by cannikin** [comment]
Went over this implementation with @mojombo and it's looking good! One issue we talked about was: we can't generate a service with a `before.skip()` in there, because now it's *not* secure by default. What do we generate in the service so that it's safe?

We came to the realization that `requireAuth` is really nice to have, but when you first build your app you'll have no authentication installed and `requireAuth` isn't defined. BUT WHAT IF IT WAS?

So we're thinking that when you generate a service, and a file is not found at `api/src/lib/auth.js`, it will be created as an absolute minimum stub:

```javascript
const requireAuth() => {
  return true
}
```

This has the benefit of no barrier when working locally and just starting out—services continue to "just work". Once you install an auth lib with `yarn rw setup auth` that file will be replaced with the provider-specific version and now all of your services are secured and you need to be logged in to access anything. You'll then need to take the step to open up those specific endpoints that are available to the world.

What if you deploy to production without any auth lib installed? Well, there's not much we can do to help in that case. If you go to production with no auth, there's really no concept of something being "private" anyway.

Any time you invoke a generator that would create a service it will double check whether `auth.js` exists and if not, create it. 

A typical generated Service will then look like:

```javascript
import { db } from 'src/lib/db'
import { requireAuth } from 'src/lib/auth'

export const beforeResolver = (before) => {
  before.run(requireAuth)
}

export const posts = () => {
  return db.post.findMany()
}

// ...
```

**Comment by cannikin** [comment]
Notes from Core meetup on 2021-04-13:

* Rename argument sent to `beforeResolver()` from `before` to `rules`
* Rename `run()` to `apply()`
* If skip() is called with no arguments, and at least one verification function resolver would be removed from the list of ones to apply, throw an error—it's super dangerous to skip EVERYTHING after setting up some logic so we'll stop you from foot-gunning yourself
* On a no-argument `skip()` add an option `force` that can be used to override that throw (proves that you REALLY thought about it and know what you're doing)

**Comment by mikebobadilla** [comment]
Just a question: Do you think an `any` type for rules might be applicable? So something that would find the first rule that returns true.

```javascript
rules.any([isContributor, isManager, isAdmin])
```

This might be a bad example, but the logic still applies.

**Comment by cannikin** [comment]
I know basically nothing about Typescript (on purpose, it's against every fiber of my being) so I have no idea where that would go or what implications it would have. 😬 I defer to @peterp!

**Comment by dac09** [comment]
I don't think he meant typescript @cannikin haha. 

I think he meant `rules.oneOf` as in, if any of these rules pass, allow the request.

**Comment by cannikin** [comment]
AHHH I get it! Using the similar `before_actions` in Rails for years I don't remember wishing I had anything like this...in this instance I'd write a single before resolver named something like `requiresPermission` that then has the conditional check for which roles are needed. 

This way you always know that each before resolver either passes or fails—there's no ambiguity when looking at the list of rules where it would be okay if *some* of them throw an error, but not others.

I don't love that added ambiguity, so I don't know that I'd use that feature, but if others find it useful...what do you think @mojombo?

**Comment by cannikin** [comment]
I'm stuck on something. We want to give people a default `auth.js` file that contains just:

```javascript
export const requireAuth = () => {
  return true
}
```

But that means when they go to setup their own auth for real, that file will always exist and the command will always fail because it doesn't want to overwrite the file. So I went down the path of looking for the `auth.js` template from `create-redwood-app` and only automatically overwrite that file if it's exactly the same (meaning they haven't touched it since they created their app from scratch).

The problem is: how do we define the files as "the same" with current state of TS vs JS along with ESLint and Prettier? The exported function could have the same behavior, but the code itself won't match because of someone's particular formatting rules, or even from deleting the comment.

So the options are:

1. Keep the default behavior: if the file exists (and it always will) the user will need to include the `--force` flag to get `yarn rw setup` to work
2. Always overwrite the `auth.js` file automatically: currently my preference as it's the smoothest experience (it just works) and if you make a mistake and overwrite something you changed, that's what git is for
3. Somehow figure out a way to determine if the files are the same: I have no idea how to do this

Again I like (2) as it seems like it would work the best with most people's behavior—creating a new app, they're not worried about auth yet, so that file likely won't be touched. If it does get significantly changed then they probably know what they're doing and won't be using our auth setup commands anyway. And even if both of those assumptions turn out to be wrong, they can roll back in git. If *all* of those assumptions are wrong, then yes they would overwrite that file and lose changes.

Preferences?

**Comment by Tobbe** [comment]
I'd like (2) more if we actually set up git by default for new projects.

I think we should go for (3)

How about this for a start?
```javascript
// auth.js
export const requireAuth = () => {
  return true
}

// auth generator
// import { requireAuth } from '...'

function isDefaultRequireAuth() {
  // Check if
  // 1. Just calling requireAuth returns true
  // 2. requireAuth doesn't take any parameters
  try {
    return requireAuth() && requireAuth.toString().slice(0, 2) === '()'
  } catch (_) {
    return false
  }
}
```

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/toString

**Comment by cannikin** [comment]
But they could have also added logic that happens to return true, or added other functions to that file, or turned it into a `default` export, or so many other things, and we'd overwrite it anyway. :(

It's a shame we didn't just make these commands interactive and ask if it's okay to overwrite. I'd be totally fine with an `api/src/lib/auth.js exists, overwrite? `y/n/a` prompt. Although since they aren't that probably involves reworking the entire generator command interface code. 😭

**Comment by Tobbe** [comment]
> But they could have also added logic that happens to return true

Of course, that's why I also added the "no parameters" check. And yes, I know it still isn't perfect.

> or turned it into a `default` export

Then the import won't work, the code will throw, and we'll return `false`

> or added other functions to that file

This is the biggest issue I think. Have no good ideas on how to handle this right now :( 

---

If you're set on just overwriting it, how about at least saving a copy of it before overwriting? `auth.js_backup`

**Comment by dac09** [comment]
Hey @cannikin just had a thought. Could we make beforeResolvers through proper 403s? So we don't have to deal with this: https://user-images.githubusercontent.com/300/104504728-c21c7680-5597-11eb-913b-3de53743dd30.png

**Comment by cannikin** [comment]
HAHAHA  Dude I'd love to, but that's not the GraphQL way is it? Shouldn't it *always* return 200, even if the datacenter is wiped out by a tsunami? 

```json
200 OK

{
  "errors": [
    {
      "message": "Datacenter no longer exists. Would you like to seek shelter?"
    }
  ]
}
```

**Comment by cannikin** [comment]
Got the prompt working!

![image](https://user-images.githubusercontent.com/300/115476605-123e5100-a1f7-11eb-8f74-28a0983e90ea.png)

* If you already passed `--force` it won't bother asking
* If you say `n` then it behaves the same as if the `--force` flag was missing (refuses to overwrite files)
* If you say `y` then it basically flips `force` to `true` for the tasks that involve overwriting files

**Comment by cannikin** [comment]
Now just need to take @dac09 up on his offer to ruin this beautiful code by adding Typescript. 😃

**Comment by dac09** [comment]
```ts
type LessMistakesInMyCodeBecauseThereAreRedSquiggles = RemoveAllTypeScriptHate<RobsMessage[]>
```

**Comment by Tobbe** [comment]
> Got the prompt working!

Nice! That was def the best option!

**Comment by cannikin** [comment]
> ```ts
> type LessMistakesInMyCodeBecauseThereAreRedSquiggles = RemoveAllTypeScriptHate<RobsMessage[]>
> ```

If I could understand this syntax I'm assuming I would be upset! 😈
